File: virtual_cat.rb
#!/usr/bin/env ruby
# virtual_cat.rb
# encoding: UTF-8

require 'json'
require 'date'
require 'fileutils'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  magenta: "\e[95m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

CAT_ASCII = <<~'CAT'
 /\_/\
( o.o )
 > ^ <
CAT

CONFIG_FILE = File.join(Dir.home, '.virtual_cat.json')

def load_state
  return { 'name' => 'Барсик', 'hunger' => 50, 'happiness' => 70,
           'energy' => 80, 'health' => 90, 'age' => 0,
           'last_update' => Time.now.iso8601 } unless File.exist?(CONFIG_FILE)
  JSON.parse(File.read(CONFIG_FILE))
end

def save_state(state)
  File.write(CONFIG_FILE, JSON.pretty_generate(state))
end

def now_iso
  Time.now.iso8601
end

def clamp(val, min=0, max=100)
  [[val, min].max, max].min
end

def get_emotion(state)
  h = state['happiness']
  return '😺' if h >= 80
  return '😸' if h >= 50
  return '😼' if h >= 30
  '😿'
end

def get_mood_text(state)
  h = state['happiness']
  return 'Мурлычет и трётся об ноги' if h >= 80
  return 'Играет с клубком ниток' if h >= 50
  return 'Сидит на подоконнике и смотрит в окно' if h >= 30
  'Лежит в углу и грустит'
end

def show_status(state)
  puts colorize(CAT_ASCII, :magenta)
  puts colorize("  Имя: #{state['name']}", :blue)
  puts colorize("  Возраст: #{state['age']} лет", :blue)
  puts "  #{get_emotion(state)}  #{get_mood_text(state)}"
  puts colorize("  Голод: #{state['hunger']}/100", state['hunger'] > 70 ? :yellow : :green)
  puts colorize("  Счастье: #{state['happiness']}/100", state['happiness'] > 50 ? :green : :red)
  puts colorize("  Энергия: #{state['energy']}/100", state['energy'] > 50 ? :green : :yellow)
  puts colorize("  Здоровье: #{state['health']}/100", state['health'] < 30 ? :red : :green)
end

def random_event(msg=nil)
  events = ['Мяу!', 'Котик трётся о ноги.', 'Принёс игрушку.',
            'Пытается поймать муху.', 'Свернулся клубком.']
  text = msg || events.sample
  puts colorize("  ✨ #{text}", :yellow)
end

def auto_mode
  puts colorize("🤖  Автоматический режим включён.", :magenta)
  loop do
    state = load_state
    state['hunger'] = clamp(state['hunger'] + rand(-5..15))
    state['happiness'] = clamp(state['happiness'] + rand(-5..10))
    state['energy'] = clamp(state['energy'] + rand(-5..10))
    state['health'] = clamp(state['health'] + rand(-3..5))
    save_state(state)
    system('clear') || system('cls')
    show_status(state)
    puts colorize("\nНажмите Ctrl+C для выхода", :yellow)
    sleep(10)
  end
rescue Interrupt
  puts colorize("\n👋  Автоматический режим завершён.", :blue)
end

def main
  if ARGV.empty?
    puts colorize("Usage: virtual_cat <status|feed|play|sleep|heal|rename|auto> [name]", :yellow)
    exit 1
  end
  action = ARGV[0]
  state = load_state

  case action
  when 'status'
    show_status(state)
  when 'feed'
    state['hunger'] = clamp(state['hunger'] - 30)
    state['happiness'] = clamp(state['happiness'] + 10)
    state['health'] = clamp(state['health'] + 5)
    puts colorize("🐟  Ням-ням! Котик поел.", :green)
    random_event("Мурлычет от удовольствия")
    save_state(state)
  when 'play'
    if state['energy'] < 20
      puts colorize("😿  Котик слишком устал для игр.", :red)
      return
    end
    state['happiness'] = clamp(state['happiness'] + 25)
    state['energy'] = clamp(state['energy'] - 20)
    state['hunger'] = clamp(state['hunger'] + 10)
    puts colorize("🧶  Игра с клубком! Котик доволен.", :green)
    random_event("Прыгает за лазерной указкой")
    save_state(state)
  when 'sleep'
    state['energy'] = clamp(state['energy'] + 40)
    state['hunger'] = clamp(state['hunger'] + 10)
    puts colorize("😴  Котик уснул. Сладких снов!", :blue)
    random_event("Во сне дёргает лапками")
    save_state(state)
  when 'heal'
    if state['energy'] < 20
      puts colorize("😿  Нет сил лечить котика.", :red)
      return
    end
    state['health'] = clamp(state['health'] + 30)
    state['energy'] = clamp(state['energy'] - 20)
    puts colorize("💊  Котик вылечен! Он благодарен.", :green)
    save_state(state)
  when 'rename'
    if ARGV.size < 2
      puts colorize("Укажите имя: rename <имя>", :red)
      return
    end
    state['name'] = ARGV[1]
    puts colorize("🐱  Котика теперь зовут #{state['name']}!", :blue)
    save_state(state)
  when 'auto'
    auto_mode
  else
    puts colorize("Неизвестное действие.", :red)
  end
end

main if __FILE__ == $0

File: virtual_cat.js
// virtual_cat.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    magenta: '\x1b[95m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const CAT_ASCII = `
 /\\_/\\
( o.o )
 > ^ <
`;

const CONFIG_FILE = path.join(os.homedir(), '.virtual_cat.json');

function loadState() {
    try {
        const data = fs.readFileSync(CONFIG_FILE, 'utf8');
        return JSON.parse(data);
    } catch (err) {
        return {
            name: 'Барсик',
            hunger: 50,
            happiness: 70,
            energy: 80,
            health: 90,
            age: 0,
            lastUpdate: new Date().toISOString()
        };
    }
}

function saveState(state) {
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(state, null, 2));
}

function nowISO() {
    return new Date().toISOString();
}

function clamp(val, min, max) {
    return Math.max(min, Math.min(val, max));
}

function getEmotion(state) {
    if (state.happiness >= 80) return '😺';
    if (state.happiness >= 50) return '😸';
    if (state.happiness >= 30) return '😼';
    return '😿';
}

function getMoodText(state) {
    if (state.happiness >= 80) return 'Мурлычет и трётся об ноги';
    if (state.happiness >= 50) return 'Играет с клубком ниток';
    if (state.happiness >= 30) return 'Сидит на подоконнике и смотрит в окно';
    return 'Лежит в углу и грустит';
}

function showStatus(state) {
    console.log(colorize(CAT_ASCII, 'magenta'));
    console.log(colorize(`  Имя: ${state.name}`, 'blue'));
    console.log(colorize(`  Возраст: ${state.age} лет`, 'blue'));
    console.log(`  ${getEmotion(state)}  ${getMoodText(state)}`);
    console.log(colorize(`  Голод: ${state.hunger}/100`, state.hunger > 70 ? 'yellow' : 'green'));
    console.log(colorize(`  Счастье: ${state.happiness}/100`, state.happiness > 50 ? 'green' : 'red'));
    console.log(colorize(`  Энергия: ${state.energy}/100`, state.energy > 50 ? 'green' : 'yellow'));
    console.log(colorize(`  Здоровье: ${state.health}/100`, state.health < 30 ? 'red' : 'green'));
}

function randomEvent(msg) {
    const events = ['Мяу!', 'Котик трётся о ноги.', 'Принёс игрушку.',
        'Пытается поймать муху.', 'Свернулся клубком.'];
    const text = msg || events[Math.floor(Math.random() * events.length)];
    console.log(colorize(`  ✨ ${text}`, 'yellow'));
}

function autoMode() {
    console.log(colorize('🤖  Автоматический режим включён.', 'magenta'));
    const interval = setInterval(() => {
        let state = loadState();
        state.hunger = clamp(state.hunger + Math.floor(Math.random() * 21) - 5, 0, 100);
        state.happiness = clamp(state.happiness + Math.floor(Math.random() * 16) - 5, 0, 100);
        state.energy = clamp(state.energy + Math.floor(Math.random() * 16) - 5, 0, 100);
        state.health = clamp(state.health + Math.floor(Math.random() * 9) - 3, 0, 100);
        saveState(state);
        console.clear();
        showStatus(state);
        console.log(colorize('\nНажмите Ctrl+C для выхода', 'yellow'));
    }, 10000);
    process.on('SIGINT', () => {
        clearInterval(interval);
        console.log(colorize('\n👋  Автоматический режим завершён.', 'blue'));
        process.exit(0);
    });
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log(colorize('Usage: node virtual_cat.js <status|feed|play|sleep|heal|rename|auto> [name]', 'yellow'));
        process.exit(1);
    }
    const action = args[0];
    const state = loadState();

    switch (action) {
        case 'status':
            showStatus(state);
            break;
        case 'feed':
            state.hunger = clamp(state.hunger - 30, 0, 100);
            state.happiness = clamp(state.happiness + 10, 0, 100);
            state.health = clamp(state.health + 5, 0, 100);
            console.log(colorize('🐟  Ням-ням! Котик поел.', 'green'));
            randomEvent('Мурлычет от удовольствия');
            saveState(state);
            break;
        case 'play':
            if (state.energy < 20) {
                console.log(colorize('😿  Котик слишком устал для игр.', 'red'));
                return;
            }
            state.happiness = clamp(state.happiness + 25, 0, 100);
            state.energy = clamp(state.energy - 20, 0, 100);
            state.hunger = clamp(state.hunger + 10, 0, 100);
            console.log(colorize('🧶  Игра с клубком! Котик доволен.', 'green'));
            randomEvent('Прыгает за лазерной указкой');
            saveState(state);
            break;
        case 'sleep':
            state.energy = clamp(state.energy + 40, 0, 100);
            state.hunger = clamp(state.hunger + 10, 0, 100);
            console.log(colorize('😴  Котик уснул. Сладких снов!', 'blue'));
            randomEvent('Во сне дёргает лапками');
            saveState(state);
            break;
        case 'heal':
            if (state.energy < 20) {
                console.log(colorize('😿  Нет сил лечить котика.', 'red'));
                return;
            }
            state.health = clamp(state.health + 30, 0, 100);
            state.energy = clamp(state.energy - 20, 0, 100);
            console.log(colorize('💊  Котик вылечен! Он благодарен.', 'green'));
            saveState(state);
            break;
        case 'rename':
            if (args.length < 2) {
                console.log(colorize('Укажите имя: rename <имя>', 'red'));
                return;
            }
            state.name = args[1];
            console.log(colorize(`🐱  Котика теперь зовут ${state.name}!`, 'blue'));
            saveState(state);
            break;
        case 'auto':
            autoMode();
            break;
        default:
            console.log(colorize('Неизвестное действие.', 'red'));
    }
}

main();
